pilot/pkg/serviceregistry/federation/controller.go

```diff
@@ -547,67 +650,107 @@ func (c *Controller) updateXDS(hostname, namespace string, instances []*model.Se
 		Name:      string(exportedService.Hostname),
-		Namespace: c.namespace,
+		Namespace: exportedService.Attributes.Namespace,
 	}] = struct{}{}
 
 	c.addServiceToStore(localService, localInstances)
 	updatedConfigs[model.ConfigKey{
 		Kind:      gvk.ServiceEntry,
 		Name:      string(localService.Hostname),
-		Namespace: c.namespace,
+		Namespace: localService.Attributes.Namespace,
 	}] = struct{}{}
 
 	return updatedConfigs, nil
 }
 
 // store has to be Lock()ed
-func (c *Controller) updateService(service *federationmodel.ServiceMessage, existing *existingImport,
-	importName federationmodel.ServiceKey) (map[model.ConfigKey]struct{}, error) {
+func (c *Controller) updateService(service *federationmodel.ServiceMessage, existing *existingImport) (map[model.ConfigKey]struct{}, error) {
+	c.logger.Debugf("handling update for exported service %+v", service.ServiceKey)
+
+	// XXX: exported service name is name.namespace, while it's namespace is c.namespace
+	importedName := c.getImportNameForService(&model.Service{
+		Hostname:   host.Name(service.Hostname),
+		Attributes: model.ServiceAttributes{Name: service.Name, Namespace: service.Namespace},
+	})
+	if importedName == nil {
+		if existing != nil {
+			c.logger.Debugf("deleting import for service %+v, as it no longer matches an import filter", service.ServiceKey)
+			return c.deleteService(service, existing), nil
+		}
+		c.logger.Debugf("skipping import for service %+v, as it does not match any import filter", service.ServiceKey)
+		return nil, nil
+	}
 
-	logger.Debugf("updating exported service %+v, known locally as %+v", service.ServiceKey, importName)
 	updatedConfigs := map[model.ConfigKey]struct{}{}
 	exportedService, exportedInstances := c.convertExportedService(service)
-	localService, localInstances := c.convertToLocalService(service, importName)
+	localService, localInstances := c.convertToLocalService(service, *importedName)
 
-	if existing != nil && importName.Hostname != existing.localName.Hostname {
-		if err := c.createRoutingResources(service.ServiceKey, importName); err != nil {
+	if existing == nil {
+		// this may have been previously filtered out
+		c.logger.Debugf("importing service %+v as %+v", service.ServiceKey, *importedName)
+		if err := c.createRoutingResources(service.ServiceKey, *importedName); err != nil {
 			return nil, err
 		}
 		// TODO: optimize service and instance updates
+	} else if importedName.Hostname != existing.localName.Hostname {
+		c.logger.Debugf("service %+v has been reimported as %+v (was %+v)", service.ServiceKey, *importedName, existing.localName)
+		// update the routing
+		if err := c.createRoutingResources(service.ServiceKey, *importedName); err != nil {
+			return nil, err
+		}
+
+		// delete the old imported service
+		c.removeServiceFromStore(existing.localName)
+		updatedConfigs[model.ConfigKey{
+			Kind:      gvk.ServiceEntry,
+			Name:      existing.localName.Hostname,
+			Namespace: existing.Namespace,
+		}] = struct{}{}
```

Comment:
Curious why not just delete the element here and below

Reply:
This is used to trigger xds, so here we're saying that the service needs to be updated (it's been deleted).  I believe this helps constrain the updates that end up being pushed out to the sidecars (i.e. so we're only pushing updates related to things that have actually changed; changed meaning added, updated, or deleted from the registry).